kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go

```diff
@@ -0,0 +1,110 @@
+package metadatacmd // import "kythe.io/kythe/go/platform/tools/kzip/metadatacmd"
+
+import (
+	"context"
+	"flag"
+	"log"
+	"time"
+
+	"kythe.io/kythe/go/platform/kzip"
+	"kythe.io/kythe/go/platform/kzip/buildmetadata"
+	"kythe.io/kythe/go/platform/tools/kzip/flags"
+	"kythe.io/kythe/go/platform/vfs"
+	"kythe.io/kythe/go/util/cmdutil"
+
+	"github.com/google/subcommands"
+)
+
+const (
+	// GitDateLayout is the default date format used by `git log`.
+	GitDateFormat = "Mon Jan 2 15:04:05 2006 -0700"
+)
+
+type metadataCommand struct {
+	cmdutil.Info
+
+	output          string
+	corpus          string
+	encoding        flags.EncodingFlag
+	commitTimestamp string
+	timestampFormat string
+}
+
+// New creates a new subcommand for creating metadata kzip files.
+func New() subcommands.Command {
+	return &metadataCommand{
+		Info:     cmdutil.NewInfo("create_metadata", "create a metadata kzip file, which records extra repository information", "--output path --commit_timestamp *"),
+		encoding: flags.EncodingFlag{Encoding: kzip.EncodingJSON},
+	}
+}
+
+// SetFlags implements the subcommands interface and provides command-specific flags
+// for merging kzip files.
+func (c *metadataCommand) SetFlags(fs *flag.FlagSet) {
+	fs.StringVar(&c.output, "output", "", "Path to output kzip file")
+	fs.StringVar(&c.corpus, "corpus", "", "Corpus to set in the unit's VName")
+	fs.StringVar(&c.commitTimestamp, "commit_timestamp", "", "Timestamp when this version of the repository was checked in")
+	fs.StringVar(&c.timestampFormat, "timestamp_format", GitDateFormat, "Format of timestamp passed to --commit_timestamp. Defaults to git's default date format: 'Mon Jan 2 15:04:05 2006 -0700'")
+	fs.Var(&c.encoding, "encoding", "Encoding to use on output, one of JSON, PROTO, or ALL")
+}
+
+// Execute implements the subcommands interface and creates a metadata kzip file.
+func (c *metadataCommand) Execute(ctx context.Context, fs *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
+	if c.output == "" {
+		return c.Fail("Required --output path missing")
+	}
+	if c.commitTimestamp == "" {
+		return c.Fail("Required --commit_timestamp missing")
+	}
+
+	t, err := time.Parse(c.timestampFormat, c.commitTimestamp)
+	if err != nil {
+		return c.Fail("Unable to parse timestamp: '%s'", c.commitTimestamp)
+	}
+	unit, err := buildmetadata.CreateMetadataUnit(t)
+	if err != nil {
+		return c.Fail("Error creating compilation unit: %v", err)
+	}
+
+	if c.corpus == "" {
+		log.Printf("WARNING: No --corpus provided")
+	}
+	unit.VName.Corpus = c.corpus
+
+	f, err := vfs.Create(ctx, c.output)
+	if err != nil {
+		return c.Fail("Creating output file: %v", err)
+	}
+	opt := kzip.WithEncoding(c.encoding.Encoding)
+	w, err := kzip.NewWriteCloser(f, opt)
+	if err != nil {
+		f.Close()
+		return c.Fail("Failed to close writer: %v", err)
+	}
+	defer w.Close()
+	if _, err := w.AddUnit(unit, nil); err != nil {
+		return c.Fail("Failed to add unit", err)
+	}
+	if err := w.Close(); err != nil {
```

Comment:
It's weird having the `defer w.Close()` and this explicit invocation.

Reply:
Agreed. Changed to just a `w.Close()` in the if statement. I could do a defer and set the return value by name, but it's kinda clunky.